Added LLM server classes (support Ollama and LM Studio)

DevNullx64 opened the pull request:
Added LLM server classes (support Ollama and LM Studio)

* Added an abstract class `LLMServerBase` with an abstract method `GetChatCompletionAsync` in `LlmServerBase.cs`.
* Implemented the `LMStudioServer` class with the `GetChatCompletionAsync` method to interact with the LM Studio server in `LmStudioServer.cs`.
* Implemented the `OllamaServer` class with the `GetChatCompletionAsync` method to interact with the Ollama server in `OllamaServer.cs`.
* Added a reference to `System.Net.Http` in the `EntwineLLM.csproj` project.
* Added a type converter (`LlmServerConverter`) for selecting the LLM server in general options (`EntwineLLMOptions.cs`).
* Modified properties `LlmUrl` and `LlmRequestTimeOut` to use values from the selected LLM server in `EntwineLLMOptions.cs`.
* Replaced direct usage of `HttpClient` with a call to the `GetChatCompletionAs`

DevNullx64 committed:
**Commit Message:**

Added LLM server classes

- Added an abstract class `LLMServerBase` with an abstract method `GetChatCompletionAsync` in `LlmServerBase.cs`.
- Implemented the `LMStudioServer` class with the `GetChatCompletionAsync` method to interact with the LM Studio server in `LmStudioServer.cs`.
- Implemented the `OllamaServer` class with the `GetChatCompletionAsync` method to interact with the Ollama server in `OllamaServer.cs`.
- Added a reference to `System.Net.Http` in the `EntwineLLM.csproj` project.
- Added a type converter (`LlmServerConverter`) for selecting the LLM server in general options (`EntwineLLMOptions.cs`).
- Modified properties `LlmUrl` and `LlmRequestTimeOut` to use values from the selected LLM server in `EntwineLLMOptions.cs`.
- Replaced direct usage of `HttpClient` with a call to the `GetChatCompletionAsync` method of the LLM server classes in `RefactoringHelper.cs`.

DevNullx64 committed:
Refactoring and Feature Addition for LLM

- Added `Microsoft.VisualStudio.Threading` import to `EntwineLLMOptions.cs`.
- Removed unnecessary imports from `EntwineLLMOptions.cs`.
- Introduced a static field `currentServer` and property `CurrentServer` in `GeneralOptions` for managing a single instance of `LLMServerBase`.
- Updated the `LlmServer` property to use `currentServer` and update it when set.
- Created `LlmModelConverter` class in `ModelsOptions` for converting LLM models using `ThreadHelper.JoinableTaskFactory`.
- Added `TypeConverter` attribute to several properties of `ModelsOptions`.
- Implemented `GetModelListAsync` method in `LlmServerBase`, `LMStudioServer`, and `OllamaServer` to retrieve model lists from HTTP endpoints.

DevNullx64 commented:
I add fetching models from Ollama or LM Studio. I've test only LM Studio ! It should work with Ollama but I've not installed it yet.

EmilianoMusso commented on src/EntwineLLM/Servers/OllamaServer.cs:
```diff
@@ -0,0 +1,44 @@
+            using var client = new HttpClient();
+            client.Timeout = RequestTimeOut;
+
+            var response = await client.GetAsync($"{BaseUrl}api/list_models");
```
I've tested it, it should be api/tags, as can be seen here: https://github.com/ollama/ollama/blob/main/docs/api.md#list-local-models

With api/list_models, it raises a 404 error, while api/tags returns the proper values

DevNullx64 replied:
Thanks !
I must install Ollama ^^'.
I will change that asap.

EmilianoMusso replied:
Thank you for your contribution, very much appreciated

EmilianoMusso commented on src/EntwineLLM/EntwineLLMOptions.cs:
```diff
@@ -1,48 +1,124 @@
 
     public class ModelsOptions : DialogPage
     {
+        public class LlmModelConverter : StringConverter
```
I think it could be better to externalise the converter class in a proper class file, to better organize the solution files

DevNullx64 replied:
When a class is only use in one place, I've the habbit to not extenalise it. But it's just a question of way to work and I agree with your opinion. 
I will update this asap.

EmilianoMusso commented on src/EntwineLLM/EntwineLLMOptions.cs:
```diff
@@ -1,48 +1,124 @@
 {
     public class GeneralOptions : DialogPage
     {
+        private class LlmServerConverter : TypeConverter
```
I think it could be better to externalise the converter class in a proper class file, to better organize the solution files

EmilianoMusso commented on src/EntwineLLM/Servers/LlmServerBase.cs:
```diff
@@ -0,0 +1,19 @@
+        public Uri BaseUrl { get; set; } = baseUri;
+        public TimeSpan RequestTimeOut { get; set; } = new TimeSpan(0, 10, 0);
+
+        public abstract Task<string[]> GetModelListAsync();
```
This method differs in LmStudioServer.cs and OllamaServer.cs by the endpoint and the property to be parsed, maybe we can unify those methods?

DevNullx64 replied:
At the current state, you fully right. But some server have spécifique implementation some time. I thinks about that another serve I wish add is Petals. But petals have not realy "chat api". It use Text Generation instead.
In real, I thin that more code need to move to LLMServerBase class. But curently I'm not confident of what to externalise.

DevNullx64 replied:
[chat.petals.dev](https://chat.petals.dev/)

EmilianoMusso replied:
Will look at it asap, thank you

EmilianoMusso replied:
> At the current state, you fully right. But some server have spécifique implementation some time. I thinks about that another serve I wish add is Petals. But petals have not realy "chat api". It use Text Generation instead. In real, I thin that more code need to move to LLMServerBase class. But curently I'm not confident of what to externalise.

Didn't know about Petals, looking at it i saw it uses a distributed LLM among users. While extremely useful perfomance-wise, i'm a little concerned about sending prompts and receiving requests from external sources, which is the reason why i wrote EntwineLLM . Let's think about the real need to include Petals

DevNullx64 replied:
Petals use HiveMind to distibute an LLM over multiple node. It's real that if one use the public swarm, the data is also public. To avoid this user can configure a private swarm. I personaly use private swarm  of Petals that connect 6 of my GPU.

Here, I wish to have more support over LLM server. An most server has some OpenAPI compatible end points and sometime other end point more powerfull. I think that have an ancestor class for LLM server inherited for specific implementation is not a bad way. Except if every server use the same API...

EmilianoMusso replied:
Ok, thank you. I think we can resolve this comment, let's keep the implementation of GetModelListAsync method on each class

DevNullx64 committed:
Refactored code and updated Ollama models API URL

- Created a new file `LlmServerConverter.cs` containing the `LlmServerConverter` class, which was previously in `EntwineLLMOptions.cs`.
- Updated the API URL in `OllamaServer.cs` from `api/list_models` to `api/tags`.

EmilianoMusso commented:
@DevNullx64 can you please update README.md and CHANGELOG.md before merging?

EmilianoMusso commented on src/EntwineLLM/LlmServerConverter.cs:
I think we can move this file in Helpers folder, updating namespace accordingly

EmilianoMusso committed:
Backmerged from master

EmilianoMusso committed:
General refactoring

kang-donghoon-eland commented:
You have done an excellent job. It has been very helpful. Thank you. Because of your assistance, I was able to connect to LM Studio.

EmilianoMusso committed:
Merged feature/general-refactor to align development
